packages/roosterjs-content-model-core/lib/coreApi/addUndoSnapshot.ts

```diff
@@ -11,14 +15,25 @@ import type { AddUndoSnapshot, Snapshot } from 'roosterjs-content-model-types';
  * when undo/redo to this snapshot
  */
 export const addUndoSnapshot: AddUndoSnapshot = (core, canUndoByBackspace, entityStates) => {
-    const { lifecycle, physicalRoot, undo } = core;
+    const { lifecycle, physicalRoot, logicalRoot, undo } = core;
     let snapshot: Snapshot | null = null;
 
     if (!lifecycle.shadowEditFragment) {
         // Need to create snapshot selection before retrieve innerHTML since HTML can be changed during creating selection when normalize table
         const selection = createSnapshotSelection(core);
         const html = physicalRoot.innerHTML;
 
+        if (!entityStates && core.physicalRoot !== core.logicalRoot) {
```

Comment:
I noticed a problem here. When we take snapshot, the logical root is a different element. Then user move logical root back to physical root, then undo. How can we know if we should change the logical root back when restore undo snapshot, and where to change to?

Reply:
Or maybe we don't need to know, and let plugin handle it?

Reply:
And I think we only need to trigger this event when logical root is under an entity (or it is entity), right? If is not under entity, the default undo behavior should be enough.

In that case, we can merge this event into EntityOperation event with a new EntityOperation type.